### PR TITLE
feat: commit title length validation

### LINKED ISSUE #31: Add lint for length of commit message

### PR DESCRIPTION
**Pull request by DeveloperC286**
Fixes https://github.com/DeveloperC286/conventional_commits_linter/issues/31

### CODE REVIEW

**Comment by DeveloperC286** [comment]
@claude got failing actions from your changes, can you check them?

**Comment by DeveloperC286** [comment]
@claude the GitHub actions checks are failing. Can you read and resolve the issues they are reporting.

**Comment by DeveloperC286** [comment]
@claude read the github actions logs and fix the compile check.

**Comment by DeveloperC286** [comment]
@claude read the github actions logs and fix the failing checks.

**Comment by DeveloperC286** [comment]
@claude read the github actions logs and fix the failing checks.

**Comment by DeveloperC286** [comment]
@claude can you run `make` to check it compiles and fix any issues if it does not.

File: end-to-end-tests/features/violations/message_too_long.feature [hunk 1]
```diff
@@ -0,0 +1,29 @@
+Feature: When the commit title exceeds the configured maximum length, it is picked up as a linting violation.
+
+
+  Scenario: Commit message that passes default length limit but fails with custom limit
```
Comment (DeveloperC286) [inline end-to-end-tests/features/violations/message_too_long.feature:4]: Change the scenarios so there is one test that starts with a too long message. You run lifting it fails then you add the arg to make it larger than 50 and it passes. The other it starts with a too small and it passes lifting, then you run it with the arg making it smaller than 50 and lifting fails.

File: src/linting_error/mod.rs [hunk 2]
```diff
@@ -17,4 +17,6 @@ pub enum LintingError {
     NoSpaceAfterColonPrecedingTypeAndScope,
     /// Commit title has no description after the type and scope.
     NoDescriptionAfterTypeAndScope,
+    /// Commit message subject line exceeds the recommended maximum length.
+    MessageTooLong,
```
Comment (DeveloperC286) [inline src/linting_error/mod.rs:21]: Can you change it to something like CommitTitleTooLong, message isn't very descriptive.

**Review by DeveloperC286 — COMMENTED** [review]
@claude actions the comments.

**Comment by DeveloperC286** [comment]
@claude actions the two new pr comments.

File: src/commits/commit/conventional_commits_specification/message_length/mod.rs [hunk 3]
```diff
@@ -0,0 +1,17 @@
+use super::*;
+
+pub fn lint(commit_message: &str, max_length: usize) -> Result<(), LintingError> {
+    // If max_length is 0, skip the length check entirely
+    if max_length == 0 {
```
Comment (DeveloperC286) [inline src/commits/commit/conventional_commits_specification/message_length/mod.rs:5]: @claude Can we convert this to an optional in main.rs, is there a way we can take an option size as the cli opt?

File: end-to-end-tests/features/steps/then-violations.py [hunk 4]
```diff
@@ -75,3 +75,12 @@ def assert_no_description_violation(context):
     assert_commits_linting_errors(result,
                                   [['NonConventionalCommitsSpecification',
                                     'NoDescriptionAfterTypeAndScope']])
+
+
+@then('a message too long violation is detected.')
```
Comment (DeveloperC286) [inline end-to-end-tests/features/steps/then-violations.py:80]: Change from message to commit title.

File: end-to-end-tests/features/violations/message_too_long.feature [hunk 5]
```diff
@@ -0,0 +1,23 @@
+Feature: When the commit title exceeds the configured maximum length, it is picked up as a linting violation.
+
+
+  Scenario: Long message fails with default limit then passes with increased limit
+    Given the context and environment are reset.
+    When linting the "feat: this is a very long commit message that exceeds fifty characters".
+    And the argument --output is set as "JSON".
+    Then a message too long violation is detected.
+    Given the context and environment are reset.
```
Comment (DeveloperC286) [inline end-to-end-tests/features/violations/message_too_long.feature:9]: No need to reset and do the commit when again.

File: end-to-end-tests/features/violations/message_too_long.feature [hunk 6]
```diff
@@ -0,0 +1,23 @@
+Feature: When the commit title exceeds the configured maximum length, it is picked up as a linting violation.
+
+
+  Scenario: Long message fails with default limit then passes with increased limit
```
Comment (DeveloperC286) [inline end-to-end-tests/features/violations/message_too_long.feature:4]: Add a full stop.

File: end-to-end-tests/features/violations/message_too_long.feature [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    Given the context and environment are reset.
+    When linting the "feat: this is a very long commit message that exceeds fifty characters".
```
Comment (DeveloperC286) [inline end-to-end-tests/features/violations/message_too_long.feature:6]: Pull the commit message out to a table like  the preceding whitespace violation feature file.

File: src/commits/commit/conventional_commits_specification/message_length/mod.rs (hunk 3 above)
Comment (DeveloperC286) [inline src/commits/commit/conventional_commits_specification/message_length/mod.rs:5]: Can we change from a size to a option<usize> if this can be disabled.

File: src/commits/commit/conventional_commits_specification/message_length/mod.rs [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+        return Ok(());
+    }
+
+    let first_line = commit_message.lines().next().unwrap_or("");
```
Comment (DeveloperC286) [inline src/commits/commit/conventional_commits_specification/message_length/mod.rs:9]: Not better to use first()?

**Comment by DeveloperC286** [comment]
@claude actions the recent comments.

File: src/commits/commit/tests/mod.rs [hunk 9]
```diff
@@ -162,3 +164,116 @@ fn test_non_conventional_commits_fail_linting(commit_message: &str) {
...
+}
+
+#[rstest(
+    commit_message,
+    case("fix: short message"),
+    case("feat: add new feature"),
+    case("docs: update docs"),
+    case("test: add unit tests for the new functionality and edge cases")
+)]
+fn test_message_within_length_limit_passes_linting(commit_message: &str) {
```
Comment (DeveloperC286) [inline src/commits/commit/tests/mod.rs:198]: Remove this unit test.

File: src/commits/commit/tests/mod.rs [hunk 10]
```diff
@@ -162,3 +164,116 @@ fn test_non_conventional_commits_fail_linting(commit_message: &str) {
 }
 
 mod generated_tests;
+
+#[rstest(
+    commit_message,
+    case("fix: This is a very long commit message that exceeds the seventy-two character limit"),
```
Comment (DeveloperC286) [inline src/commits/commit/tests/mod.rs:170]: Delete this unit test but move one example into the existing failing_linting unit test.

File: end-to-end-tests/features/violations/message_too_long.feature [hunk 11]
```diff
@@ -0,0 +1,29 @@
+Feature: When the commit title exceeds the configured maximum length, it is picked up as a linting violation.
+
+
+  Scenario Outline: Long commit title fails with default limit then passes with increased limit
+    Given the context and environment are reset.
+    When linting the "<commit_message>".
+    And the argument --output is set as "JSON".
+    Then a commit title too long violation is detected.
+    When linting the "<commit_message>".
```
Comment (DeveloperC286) [inline end-to-end-tests/features/violations/message_too_long.feature:9]: No need for this when.

File: end-to-end-tests/features/violations/message_too_long.feature [hunk 12]
```diff
@@ -0,0 +1,29 @@
+Feature: When the commit title exceeds the configured maximum length, it is picked up as a linting violation.
+
+
+  Scenario Outline: Long commit title fails with default limit then passes with increased limit
```
Comment (DeveloperC286) [inline end-to-end-tests/features/violations/message_too_long.feature:4]: Add a fullstop.

File: end-to-end-tests/features/violations/message_too_long.feature [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+    Examples:
+      | commit_message                                                    |
+      | feat: this is a very long commit message that exceeds fifty characters |
+
+
+  Scenario Outline: Short commit title passes with default limit then fails with decreased limit
+    Given the context and environment are reset.
+    When linting the "<commit_message>".
+    Then the linting passes.
+    When linting the "<commit_message>".
```
Comment (DeveloperC286) [inline end-to-end-tests/features/violations/message_too_long.feature:22]: No need for this when.

File: end-to-end-tests/features/violations/message_too_long.feature [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+    And the argument --max-commit-title-length is set to "80".
+    Then the linting passes.
+
+    Examples:
+      | commit_message                                                    |
+      | feat: this is a very long commit message that exceeds fifty characters |
+
+
+  Scenario Outline: Short commit title passes with default limit then fails with decreased limit
```
Comment (DeveloperC286) [inline end-to-end-tests/features/violations/message_too_long.feature:18]: Add a fullstop.

File: src/commits/commit/conventional_commits_specification/message_length/mod.rs [hunk 15]
```diff
@@ -0,0 +1,18 @@
+use super::*;
+
+pub fn lint(commit_message: &str, max_length: Option<usize>) -> Result<(), LintingError> {
+    // If max_length is None, skip the length check entirely
```
Comment (DeveloperC286) [inline src/commits/commit/conventional_commits_specification/message_length/mod.rs:4]: Useless comment.

File: src/commits/commit/mod.rs [hunk 16]
```diff
@@ -106,6 +106,19 @@ impl Commit {
             }
         }
 
+        // Check message length regardless of conventional commits compliance
+        let max_length = if max_commit_title_length == 0 {
```
Comment (DeveloperC286) [inline src/commits/commit/mod.rs:110]: Should this logic not be in main.rs where the cli logic is.

**Comment by DeveloperC286** [comment]
@claude action all new comments.

File: src/commits/commit/conventional_commits_specification/message_length/mod.rs [hunk 17]
```diff
@@ -0,0 +1,17 @@
+use super::*;
+
+pub fn lint(commit_message: &str, max_length: Option<usize>) -> Result<(), LintingError> {
+    let max_length = match max_length {
+        Some(length) => length,
+        None => return Ok(()),
+    };
+
+    let first_line = commit_message.lines().next().ok_or(LintingError::CommitTitleTooLong)?;
+    let subject_length = first_line.chars().count();
```
Comment (DeveloperC286) [inline src/commits/commit/conventional_commits_specification/message_length/mod.rs:10]: Call it length.

File: src/commits/commit/conventional_commits_specification/message_length/mod.rs [hunk 18]
```diff
@@ -0,0 +1,17 @@
+use super::*;
+
+pub fn lint(commit_message: &str, max_length: Option<usize>) -> Result<(), LintingError> {
+    let max_length = match max_length {
+        Some(length) => length,
+        None => return Ok(()),
+    };
+
+    let first_line = commit_message.lines().next().ok_or(LintingError::CommitTitleTooLong)?;
```
Comment (DeveloperC286) [inline src/commits/commit/conventional_commits_specification/message_length/mod.rs:9]: Call it title and just call unwrap or something for now,  no default.

File: src/commits/commit/conventional_commits_specification/message_length/mod.rs [hunk 19]
```diff
@@ -0,0 +1,17 @@
+use super::*;
+
+pub fn lint(commit_message: &str, max_length: Option<usize>) -> Result<(), LintingError> {
+    let max_length = match max_length {
```
Comment (DeveloperC286) [inline src/commits/commit/conventional_commits_specification/message_length/mod.rs:4]: Call it max and maybe just put the logic inside the some branch?

File: src/commits/commit/tests/mod.rs [hunk 20]
```diff
@@ -162,3 +165,72 @@ fn test_non_conventional_commits_fail_linting(commit_message: &str) {
...
+fn test_message_too_long_with_custom_limit_fails_linting(commit_message: &str) {
+    // Given
+    let commit = Commit::from_commit_message(commit_message.to_string());
+
+    // When - using a custom limit of 50 characters
+    let linting_errors = commit.lint(false, 50);
+
+    // Then
+    assert!(
+        linting_errors.contains(&LintingError::CommitTitleTooLong),
```
Comment (DeveloperC286) [inline src/commits/commit/tests/mod.rs:186]: Add a too long variation to the generated tests.

**Comment by DeveloperC286** [comment]
@claude action the new comments.
